streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java

```diff
@@ -198,7 +207,8 @@ public <K, V1, V2, VOut> KStream<K, VOut> join(final KStream<K, V1> lhs,
                    .withOtherWindowedStreamProcessorParameters(otherWindowStreamProcessorParams)
                    .withOuterJoinWindowStoreBuilder(outerJoinWindowStore)
                    .withValueJoiner(joiner)
-                   .withNodeName(joinMergeName);
+                   .withNodeName(joinMergeName)
+                   .withSelfJoinProcessorParameters(selfJoinProcessorParams);
```

Comment:
Not a strong opinion, since I know we are messing the logical planner with physical info quite badly already.. but I'm wondering if we could defer this in `StreamStreamJoinNode#writeToTopology` inside the `isSelfJoin` condition? We still have all the pieces we need: 1) left store name, 2) both join window specs. 3) joiner, 4) time tracker, from other params, so that in `writeToTopology` we can still generate the `KStreamKStreamSelfJoin` object if self-join is enabled.

My motivation is just to not spill more physical node info into logical planning phase.

Reply:
Hey @guozhangwang! You mean to move to `StreamStreamJoinNode#writeToTopology`  the code about the creation of the self-join node 
```
final KStreamKStreamSelfJoin<K, V1, V2, VOut> selfJoin = new KStreamKStreamSelfJoin<>(
            thisWindowStore.name(),
            internalWindows,
            joiner,
            sharedTimeTracker
        );
```
I tried it but I don't think it makes the code clearer. All the components of the `StreamStreamJoinNode` are built in the `KStreamJoinImpl` and are only visible there because they are not `public`. So, I would have to special case the self-join part which fixes the problem of spilling physical information into the logical plan but does it only for a small special case. I am worried this will be confusing.

Reply:
Ack, thanks a lot for letting me know :)